cmd/client_test.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2020 Google LLC // // Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Evan Spendlove (@evanSpendlove)
+//
+// Client_test tests the client.go file to ensure that the gRPC client for connecting
+// to Cloudprober works as expected. This client is used for adding and removing probes.
+
+package cmd
+
+import (
+	"github.com/golang/glog"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/google/cloudprober"
+	"github.com/google/cloudprober/examples/extensions/myprober/myprobe"
+	probes_configpb "github.com/google/cloudprober/probes/proto"
+	targetspb "github.com/google/cloudprober/targets/proto"
+	"github.com/googleinterns/step224-2020/hermes"
+)
+
+const (
+	cpCfg     string = "grpc_port:9314" // Cloudprober config
+	rpcServer string = "localhost:9314" // Cloudprober gRPC server address
+)
+
+// TODO: Make these proper table-driven tests.
+var (
+	testProbes = []string{
+		"testExtension0",
+		"testExtension1"}
+)
+
+func setupCloudproberAndClient() (*hermes.Hermes, *CloudproberClient) {
+	testHermes := &hermes.Hermes{}
+	err := testHermes.InitialiseCloudproberFromConfig(cpCfg)
+	if err != nil {
+		glog.Fatalf("Cloudprober could not be initialised, err: %v", err)
+	}
+
+	cloudprober.Start(testHermes.Ctx)
+
+	client, errClient := NewClient(rpcServer)
+	if errClient != nil {
+		glog.Fatalf("Cloudprober gRPC Client could not be initialised: %v", errClient)
+	}
+
+	return testHermes, client
+}
+
+func teardownCloudproberAndClient(testHermes *hermes.Hermes, client *CloudproberClient) {
+	// TODO: Find a safe way to close down Cloudprober and restart it.
+	for i := 0; i < len(testProbes); i++ {
+		err := client.RemoveProbe(testProbes[i])
+		if err != nil {
+			glog.Errorf("RemoveProbe() failed, expected error: %v, got %v", nil, err)
+		}
+
+	}
+
+	client.CloseConn()
+
+	// Need to wait for the RemoveProbe() RPCs to be processed on the server side
+	time.Sleep(1 * time.Second)
```

Comment:
The problem isn't the locking on the Hermes side, but when I remove this, it seems that `RemoveProbe()` RPCs are sent, but not fully processed on the Cloudprober side before the `ListProbes()` RPC is sent to check which probes are currently running.

This problem only occurs approx. once every 10-15 runs of the tests, so it is quite difficult to debug.

I have verified the output of client.CloseConn(), there is no error thrown here.

Reply:
I'm not 100% sure what is causing this issue as Cloudprober's logs are not detailed enough, and the gRPC logs just show me that the connection is made successfully, etc.

If you have any thoughts/insight on this, I would really appreciate them!!!

Reply:
Using sleep to fix test flakiness is a code smell.
It makes the tests run slower and sweeps the problem under the rug. It also doens't guarantee that flakiness is eliminated, as under heavy load even the one second delay might not be enough.
I would recommend adding explicit synchronization, even if it's just for testing.

Reply:
I think that I have resolved the core of this issue now, so I have removed the `Sleep()`.

The tests failed because `ListProbes()` does not return a sorted list, so I have added sorting to the relevant tests, and this error has not occurred since.